geopyspark-backend/geotrellis/src/main/scala/geopyspark/geotrellis/vlm/RasterSource.scala

```diff
@@ -0,0 +1,158 @@
+import geopyspark.geotrellis.{LayoutType => GPSLayoutType, LocalLayout => GPSLocalLayout, GlobalLayout => GPSGlobalLayout}
+
+import geopyspark.geotrellis.Constants.{GEOTRELLIS, GDAL}
+
+import geotrellis.contrib.vlm._
+import geotrellis.contrib.vlm.geotiff.GeoTiffRasterSource
+import geotrellis.contrib.vlm.gdal.GDALRasterSource
+
+import geotrellis.raster._
+import geotrellis.raster.resample._
+import geotrellis.spark._
+import geotrellis.spark.tiling._
+import geotrellis.proj4._
+import geotrellis.vector._
+import geotrellis.util._
+
+import org.apache.spark.SparkContext
+import org.apache.spark.rdd.RDD
+
+import scala.collection.JavaConverters._
+
+
+object RasterSource {
+  def read(
+    sc: SparkContext,
+    layerType: String,
+    paths: java.util.ArrayList[String],
+    targetCRS: String,
+    resampleMethod: ResampleMethod,
+    readMethod: String
+  ): ProjectedRasterLayer =
+    read(
+      sc,
+      layerType,
+      sc.parallelize(paths.asScala),
+      targetCRS,
+      resampleMethod,
+      readMethod
+    )
+
+  def read(
+    sc: SparkContext,
+    layerType: String,
+    rdd: RDD[String],
+    targetCRS: String,
+    resampleMethod: ResampleMethod,
+    readMethod: String
+  ): ProjectedRasterLayer = {
+    val rasterSourceRDD: RDD[RasterSource] =
+      (readMethod match {
+        case GEOTRELLIS => rdd.map { new GeoTiffRasterSource(_): RasterSource }
+        case GDAL => rdd.map { GDALRasterSource(_): RasterSource }
+      }).cache()
+
+    val reprojectedSourcesRDD: RDD[RasterSource] =
+      targetCRS match {
+        case crs: String =>
+          rasterSourceRDD.map { _.reproject(CRS.fromString(crs), resampleMethod) }
+        case null =>
+          rasterSourceRDD
+      }
+
+    val projectedRasterRDD: RDD[(ProjectedExtent, MultibandTile)] =
+      reprojectedSourcesRDD.flatMap { source: RasterSource =>
+        source.read(source.extent) match {
+          case Some(raster) => Some((ProjectedExtent(raster.extent, source.crs), raster.tile))
+          case None => None
+        }
+      }
+
+    rasterSourceRDD.unpersist()
+
+    ProjectedRasterLayer(projectedRasterRDD)
+  }
+
+  def readToLayout(
+    sc: SparkContext,
+    layerType: String,
+    paths: java.util.ArrayList[String],
+    layoutType: GPSLayoutType,
+    targetCRS: String,
+    resampleMethod: ResampleMethod,
+    readMethod: String
+  ): SpatialTiledRasterLayer =
+    readToLayout(
+      sc,
+      layerType,
+      sc.parallelize(paths.asScala),
+      layoutType,
+      targetCRS,
+      resampleMethod,
+      readMethod
+    )
+
+  def readToLayout(
+    sc: SparkContext,
+    layerType: String,
+    rdd: RDD[String],
+    layoutType: LayoutType,
+    targetCRS: String,
+    resampleMethod: ResampleMethod,
+    readMethod: String
+  ): SpatialTiledRasterLayer = {
+    // TODO: These are the things that still need to be done:
+    // 1. Support TemporalTiledRasterLayer (ie. generic K)
+    // 2. Use the partitionStrategy parameter
+
```

Comment:
Needs an extra TODO to use `summary.estimatePartitionsNumber `
ex: https://github.com/geotrellis/geotrellis-landsat-tutorial/blob/feature/raster-source/src/main/scala/tutorial/vlm/RasterSourceLocalIngestImage.scala#L60

I'm not sure how big of a deal it is but potentially current method will create very unbalanced partitions.

`sc.parallelize(paths.asScala)` is going to default to `spark.default.parallelism` which I believe defaults for `4` or something like that.

Reply:
Are you planning to get that in as part of band multiplex feature or do you think it can be an outstanding issue for a bit ?

Reply:
I was going to include a `numPartitions` parameter in the multiplex PR as a temporary solution. Unless we're going to be using this feature soon, it can probably be an outstanding issue. Though, it's something we should address sooner than later.

Reply:
Ideally this becomes the way you read all raster in GPS, so it would be worth while to take care of it then.

Reply:
Agreed. I think it's something we should do next sprint.